dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -318,32 +371,313 @@ private List<Event> inlineAtomicEnd(FunctionCall ignored) {
+        return List.of(
+                EventFactory.newLoad(result, expressions.makeADD(key, offset)));
+    }
+
+    private List<Event> inlinePthreadSetSpecific(FunctionCall call) {
+        final Register result = checkValueAndArguments(2, call);
+        final Expression key = call.getArguments().get(0);
+        final Expression value = call.getArguments().get(1);
+        final int threadID = call.getThread().getId();
+        final Expression offset = expressions.makeValue(BigInteger.valueOf(threadID), types.getArchType());
+        return List.of(
+                EventFactory.newLocal(result, expressions.makeGeneralZero(result.getType())),
+                EventFactory.newStore(expressions.makeADD(key, offset), value));
+    }
+
     private List<Event> inlinePthreadMutexInit(FunctionCall call) {
-        checkArgument(call.getArguments().size() == 2);
+        final Register result = checkValueAndArguments(2, call);
         final Expression lockAddress = call.getArguments().get(0);
-        final Expression lockValue = call.getArguments().get(1);
+        final Expression attributes = call.getArguments().get(1);
         final String lockName = lockAddress.toString();
-        return List.of(EventFactory.Pthread.newInitLock(lockName, lockAddress, lockValue));
+        return List.of(
+                EventFactory.Pthread.newInitLock(lockName, lockAddress, attributes),
+                EventFactory.newLocal(result, expressions.makeGeneralZero(result.getType())));
     }
 
     private List<Event> inlinePthreadMutexDestroy(FunctionCall call) {
-        checkArgument(call.getArguments().size() == 1);
-        final Register reg = ((ValueFunctionCall) call).getResultRegister();
-        return List.of(EventFactory.newLocal(reg, expressions.makeZero((IntegerType) reg.getType())));
+        final Register result = checkValueAndArguments(1, call);
+        return List.of(
+                EventFactory.newLocal(result, expressions.makeGeneralZero(result.getType())));
     }
 
     private List<Event> inlinePthreadMutexLock(FunctionCall call) {
-        checkArgument(call.getArguments().size() == 1);
+        final Register result = checkValueAndArguments(1, call);
         final Expression lockAddress = call.getArguments().get(0);
         final String lockName = lockAddress.toString();
-        return List.of(EventFactory.Pthread.newLock(lockName, lockAddress));
+        return List.of(
+                EventFactory.Pthread.newLock(lockName, lockAddress),
+                EventFactory.newLocal(result, expressions.makeGeneralZero(result.getType())));
+    }
+
+    private List<Event> inlinePthreadMutexTryLock(FunctionCall call) {
+        final Register register = checkValueAndArguments(1, call);
+        checkArgument(register.getType() instanceof IntegerType, "Wrong return type for \"%s\"", call);
+        final var error = new INonDet(constantId++, (IntegerType) register.getType(), true);
```

Comment:
Yes, this is a bug.  I did not want to insert a CAS here, since I did not know if it would have interfered with `Lock` and `Unlock`.  Instead I could add a new `TryLock` class or replace the existing lock events with atomics.

Reply:
Btw. I think if we have like 80% of the pthread library implemented via C11 atomics, we can also go all the way and do so for the standard lock, unlock, and init, that are currently handled by special events. 
We are unlikely to do verification w.r.t. to the full C11 model in the near future and that is the only model that talks about these locks as special events. It is also unlikely that the standard implementation via acquire/release deviates much from the ordering implied by the C11 model (so we are unlikely to miss bugs with a concrete implementation).
Lastly, even if we decide to do full C11, we have to introduce a lot of new events anyhow, so adding back in the three we kicked out is not going to be a big deal.